File: client/src/lib/server/tool-registry.ts
/**
 * Tool registry — data-driven lookup table for "Open In" tools.
 *
 * Replaces the hardcoded switch statement in open-directory/route.ts
 * with an extensible, platform-aware registry of editors, terminals,
 * and file explorers.
 */

import { resolve } from "path";

// ── Types ───────────────────────────────────────────────────────────────────

export type ToolCategory = "editor" | "terminal" | "explorer";

export type PlatformId = "win32" | "darwin" | "linux";

export interface PlatformCommand {
  /** The binary or command to execute. */
  command: string;
  /** Build the args array for a given directory path. */
  args: (dir: string) => string[];
  /** Optional spawn options overrides. */
  options?: {
    shell?: boolean;
    /** When set to "directory", the spawn cwd is set to the target directory. */
    cwd?: "directory";
    windowsHide?: boolean;
  };
}

export interface ToolDefinition {
  id: string;
  label: string;
  /** Lucide icon name string — mapped to a component client-side. */
  iconName: string;
  category: ToolCategory;
  /** Per-platform spawn configuration. Absent key = tool not available on that platform. */
  platforms: Partial<Record<PlatformId, PlatformCommand>>;
  /** If true, the tool is always considered available (no detection needed). */
  alwaysAvailable?: boolean;
  /**
   * The binary names to probe during detection.
   * If omitted the detector derives it from the platform command.
   */
  detectBinaries?: Partial<Record<PlatformId, string[]>>;
  /** macOS .app bundle names to check in /Applications. */
  detectMacApps?: string[];
}

// ── Builtin tool definitions ────────────────────────────────────────────────

export const BUILTIN_TOOLS: readonly ToolDefinition[] = [
  // ── Editors ───────────────────────────────────────────────────────────────
  {
    id: "vscode",
    label: "VS Code",
    iconName: "code-2",
    category: "editor",
    platforms: {
      darwin: {
        command: "open",
        args: (dir) => ["-a", "Visual Studio Code", dir],
      },
      win32: {
        command: "code",
        args: (dir) => [dir],
        options: { shell: true, windowsHide: true },
      },
      linux: {
        command: "code",
        args: (dir) => [dir],
      },
    },
    detectBinaries: {
      win32: ["code"],
      linux: ["code"],
    },
    detectMacApps: ["Visual Studio Code"],
  },
  {
    id: "vscode-insiders",
    label: "VS Code Insiders",
    iconName: "code-2",
    category: "editor",
    platforms: {
      darwin: {
        command: "open",
        args: (dir) => ["-a", "Visual Studio Code - Insiders", dir],
      },
      win32: {
        command: "code-insiders",
        args: (dir) => [dir],
        options: { shell: true, windowsHide: true },
      },
      linux: {
        command: "code-insiders",
        args: (dir) => [dir],
      },
    },
    detectBinaries: {
      win32: ["code-insiders"],
      linux: ["code-insiders"],
    },
    detectMacApps: ["Visual Studio Code - Insiders"],
  },
  {
    id: "cursor",
    label: "Cursor",
    iconName: "mouse-pointer-2",
    category: "editor",
    platforms: {
      darwin: {
        command: "open",
        args: (dir) => ["-a", "Cursor", dir],
      },
      win32: {
        command: "cursor",
        args: (dir) => [dir],
        options: { shell: true, windowsHide: true },
      },
      linux: {
        command: "cursor",
        args: (dir) => [dir],
      },
    },
    detectBinaries: {
      win32: ["cursor"],
      linux: ["cursor"],
    },
    detectMacApps: ["Cursor"],
  },
  {
    id: "windsurf",
    label: "Windsurf",
    iconName: "wind",
    category: "editor",
    platforms: {
      darwin: {
        command: "open",
        args: (dir) => ["-a", "Windsurf", dir],
      },
      win32: {
        command: "windsurf",
        args: (dir) => [dir],
        options: { shell: true, windowsHide: true },
      },
      linux: {
        command: "windsurf",
        args: (dir) => [dir],
      },
    },
    detectBinaries: {
      win32: ["windsurf"],
      linux: ["windsurf"],
    },
    detectMacApps: ["Windsurf"],
  },
  {
    id: "zed",
    label: "Zed",
    iconName: "pen-tool",
    category: "editor",
    platforms: {
      darwin: {
        command: "open",
        args: (dir) => ["-a", "Zed", dir],
      },
      linux: {
        command: "zed",
        args: (dir) => [dir],
      },
    },
    detectBinaries: {
      linux: ["zed"],
    },
    detectMacApps: ["Zed"],
  },
  {
    id: "sublime",
    label: "Sublime Text",
    iconName: "braces",
    category: "editor",
    platforms: {
      darwin: {
        command: "open",
        args: (dir) => ["-a", "Sublime Text", dir],
      },
      win32: {
        command: "subl",
        args: (dir) => [dir],
        options: { shell: true, windowsHide: true },
      },
      linux: {
        command: "subl",
        args: (dir) => [dir],
      },
    },
    detectBinaries: {
      win32: ["subl"],
      linux: ["subl"],
    },
    detectMacApps: ["Sublime Text"],
  },
  {
    id: "neovim",
    label: "Neovim",
    iconName: "square-terminal",
    category: "editor",
    platforms: {
      darwin: {
        command: "nvim",
        args: (dir) => [dir],
      },
      win32: {
        command: "nvim",
        args: (dir) => [dir],
        options: { shell: true, windowsHide: true },
      },
      linux: {
        command: "nvim",
        args: (dir) => [dir],
      },
    },
    detectBinaries: {
      darwin: ["nvim"],
      win32: ["nvim"],
      linux: ["nvim"],
    },
  },
  {
    id: "intellij",
    label: "IntelliJ IDEA",
    iconName: "app-window",
    category: "editor",
    platforms: {
      darwin: {
        command: "open",
        args: (dir) => ["-a", "IntelliJ IDEA", dir],
      },
      win32: {
        command: "idea",
        args: (dir) => [dir],
        options: { shell: true, windowsHide: true },
      },
      linux: {
        command: "idea",
        args: (dir) => [dir],
      },
    },
    detectBinaries: {
      win32: ["idea", "idea64"],
      linux: ["idea"],
    },
    detectMacApps: ["IntelliJ IDEA", "IntelliJ IDEA CE"],
  },
  {
    id: "webstorm",
    label: "WebStorm",
    iconName: "app-window",
    category: "editor",
    platforms: {
      darwin: {
        command: "open",
        args: (dir) => ["-a", "WebStorm", dir],
      },
      win32: {
        command: "webstorm",
        args: (dir) => [dir],
        options: { shell: true, windowsHide: true },
      },
      linux: {
        command: "webstorm",
        args: (dir) => [dir],
      },
    },
    detectBinaries: {
      win32: ["webstorm", "webstorm64"],
      linux: ["webstorm"],
    },
    detectMacApps: ["WebStorm"],
  },
  {
    id: "rider",
    label: "Rider",
    iconName: "app-window",
    category: "editor",
    platforms: {
      darwin: {
        command: "open",
        args: (dir) => ["-a", "Rider", dir],
      },
      win32: {
        command: "rider",
        args: (dir) => [dir],
        options: { shell: true, windowsHide: true },
      },
      linux: {
        command: "rider",
        args: (dir) => [dir],
      },
    },
    detectBinaries: {
      win32: ["rider", "rider64"],
      linux: ["rider"],
    },
    detectMacApps: ["Rider"],
  },
  {
    id: "goland",
    label: "GoLand",
    iconName: "app-window",
    category: "editor",
    platforms: {
      darwin: {
        command: "open",
        args: (dir) => ["-a", "GoLand", dir],
      },
      win32: {
        command: "goland",
        args: (dir) => [dir],
        options: { shell: true, windowsHide: true },
      },
      linux: {
        command: "goland",
        args: (dir) => [dir],
      },
    },
    detectBinaries: {
      win32: ["goland", "goland64"],
      linux: ["goland"],
    },
    detectMacApps: ["GoLand"],
  },
  {
    id: "pycharm",
    label: "PyCharm",
    iconName: "app-window",
    category: "editor",
    platforms: {
      darwin: {
        command: "open",
        args: (dir) => ["-a", "PyCharm", dir],
      },
      win32: {
        command: "pycharm",
        args: (dir) => [dir],
        options: { shell: true, windowsHide: true },
      },
      linux: {
        command: "pycharm",
        args: (dir) => [dir],
      },
    },
    detectBinaries: {
      win32: ["pycharm", "pycharm64"],
      linux: ["pycharm"],
    },
    detectMacApps: ["PyCharm", "PyCharm CE"],
  },
  {
    id: "rustrover",
    label: "RustRover",
    iconName: "app-window",
    category: "editor",
    platforms: {
      darwin: {
        command: "open",
        args: (dir) => ["-a", "RustRover", dir],
      },
      win32: {
        command: "rustrover",
        args: (dir) => [dir],
        options: { shell: true, windowsHide: true },
      },
      linux: {
        command: "rustrover",
        args: (dir) => [dir],
      },
    },
    detectBinaries: {
      win32: ["rustrover", "rustrover64"],
      linux: ["rustrover"],
    },
    detectMacApps: ["RustRover"],
  },
  {
    id: "fleet-jb",
    label: "Fleet (JetBrains)",
    iconName: "app-window",
    category: "editor",
    platforms: {
      darwin: {
        command: "open",
        args: (dir) => ["-a", "Fleet", dir],
      },
      win32: {
        command: "fleet",
        args: (dir) => [dir],
        options: { shell: true, windowsHide: true },
      },
      linux: {
        command: "fleet",
        args: (dir) => [dir],
      },
    },
    detectBinaries: {
      win32: ["fleet"],
      linux: ["fleet"],
    },
    detectMacApps: ["Fleet"],
  },

  // ── Terminals ─────────────────────────────────────────────────────────────
  {
    id: "terminal",
    label: "System Terminal",
    iconName: "terminal",
    category: "terminal",
    alwaysAvailable: true,
    platforms: {
      darwin: {
        command: "open",
        args: () => ["-a", "Terminal", "."],
        options: { cwd: "directory" },
      },
      win32: {
        command: "cmd",
        args: () => ["/c", "start", "cmd", "/K"],
        options: { shell: true, cwd: "directory" },
      },
      linux: {
        command: "x-terminal-emulator",
        args: () => [],
        options: { cwd: "directory" },
      },
    },
  },
  {
    id: "wt",
    label: "Windows Terminal",
    iconName: "square-terminal",
    category: "terminal",
    platforms: {
      win32: {
        command: "wt",
        args: (dir) => ["-d", dir],
        options: { shell: true, windowsHide: true },
      },
    },
    detectBinaries: {
      win32: ["wt"],
    },
  },
  {
    id: "iterm2",
    label: "iTerm2",
    iconName: "square-terminal",
    category: "terminal",
    platforms: {
      darwin: {
        command: "open",
        args: (dir) => ["-a", "iTerm", dir],
      },
    },
    detectMacApps: ["iTerm"],
  },
  {
    id: "alacritty",
    label: "Alacritty",
    iconName: "square-terminal",
    category: "terminal",
    platforms: {
      darwin: {
        command: "alacritty",
        args: (dir) => ["--working-directory", dir],
      },
      win32: {
        command: "alacritty",
        args: (dir) => ["--working-directory", dir],
        options: { shell: true, windowsHide: true },
      },
      linux: {
        command: "alacritty",
        args: (dir) => ["--working-directory", dir],
      },
    },
    detectBinaries: {
      darwin: ["alacritty"],
      win32: ["alacritty"],
      linux: ["alacritty"],
    },
  },
  {
    id: "warp",
    label: "Warp",
    iconName: "square-terminal",
    category: "terminal",
    platforms: {
      darwin: {
        command: "open",
        args: (dir) => ["-a", "Warp", dir],
      },
      linux: {
        command: "warp-terminal",
        args: (dir) => [dir],
      },
    },
    detectBinaries: {
      linux: ["warp-terminal"],
    },
    detectMacApps: ["Warp"],
  },
  {
    id: "hyper",
    label: "Hyper",
    iconName: "square-terminal",
    category: "terminal",
    platforms: {
      darwin: {
        command: "open",
        args: (dir) => ["-a", "Hyper", dir],
      },
      win32: {
        command: "hyper",
        args: (dir) => [dir],
        options: { shell: true, windowsHide: true },
      },
      linux: {
        command: "hyper",
        args: (dir) => [dir],
      },
    },
    detectBinaries: {
      win32: ["hyper"],
      linux: ["hyper"],
    },
    detectMacApps: ["Hyper"],
  },
  {
    id: "kitty",
    label: "kitty",
    iconName: "square-terminal",
    category: "terminal",
    platforms: {
      darwin: {
        command: "kitty",
        args: (dir) => ["--directory", dir],
      },
      linux: {
        command: "kitty",
        args: (dir) => ["--directory", dir],
      },
    },
    detectBinaries: {
      darwin: ["kitty"],
      linux: ["kitty"],
    },
  },

  // ── File Explorers ────────────────────────────────────────────────────────
  {
    id: "explorer",
    label: "File Explorer",
    iconName: "folder-open",
    category: "explorer",
    alwaysAvailable: true,
    platforms: {
      darwin: {
        command: "open",
        args: (dir) => [dir],
      },
      win32: {
        command: "explorer",
        args: (dir) => [dir],
      },
      linux: {
        command: "xdg-open",
        args: (dir) => [dir],
      },
    },
  },
];

// ── Lookup helpers ──────────────────────────────────────────────────────────

const toolIndex = new Map<string, ToolDefinition>(
  BUILTIN_TOOLS.map((t) => [t.id, t])
);

/** Look up a builtin tool by ID. */
export function getToolById(id: string): ToolDefinition | undefined {
  return toolIndex.get(id);
}

/** Get all builtin tools in a given category. */
export function getToolsByCategory(category: ToolCategory): ToolDefinition[] {
  return BUILTIN_TOOLS.filter((t) => t.category === category);
}

/** Return every known builtin tool ID. */
export function getAllToolIds(): string[] {
  return BUILTIN_TOOLS.map((t) => t.id);
}

/** Check if an ID matches a builtin tool. */
export function isValidToolId(id: string): boolean {
  return toolIndex.has(id);
}

// ── Config-driven resolution ────────────────────────────────────────────────

export interface ResolvedTool {
  id: string;
  label: string;
  iconName: string;
  category: ToolCategory;
}

export interface WeaveToolOverride {
  command?: string;
  args?: string;
  hidden?: boolean;
}

export interface WeaveCustomTool {
  label: string;
  category: ToolCategory;
  iconName?: string;
  command: string;
  args?: string;
  platforms?: PlatformId[];
}

export interface WeaveToolsConfig {
  overrides?: Record<string, WeaveToolOverride>;
  custom?: Record<string, WeaveCustomTool>;
}

/**
 * Merge detected tools with config overrides and custom tool definitions.
 * Returns the final list of tools to show in the UI.
 */
export function resolveTools(
  detected: ToolDefinition[],
  config: WeaveToolsConfig | undefined
): ResolvedTool[] {
  const overrides = config?.overrides ?? {};
  const custom = config?.custom ?? {};
  const platform = process.platform as PlatformId;

  // Start with detected builtins, applying overrides
  const resolved: ResolvedTool[] = [];
  for (const tool of detected) {
    const override = overrides[tool.id];
    if (override?.hidden) continue;
    resolved.push({
      id: tool.id,
      label: tool.label,
      iconName: tool.iconName,
      category: tool.category,
    });
  }

  // Append custom tools (these skip detection — user guarantees they exist)
  for (const [id, def] of Object.entries(custom)) {
    if (overrides[id]?.hidden) continue;
    // Skip if restricted to other platforms
    if (def.platforms && !def.platforms.includes(platform)) continue;
    resolved.push({
      id,
      label: def.label,
      iconName: def.iconName ?? "wrench",
      category: def.category,
    });
  }

  return resolved;
}

/**
 * Split an args template string into an array, preserving `${dir}` as a
 * single token even when the directory path contains spaces.
 *
 * The template is first split on whitespace, then each token has `${dir}`
 * replaced with the actual directory path.  Because the split happens
 * *before* substitution, spaces inside the directory value never cause
 * an argument to be broken apart.
 */
function expandArgsTemplate(template: string, directory: string): string[] {
  return template
    .split(/\s+/)
    .filter(Boolean)
    .map((token) => token.replace(/\$\{dir\}/g, directory));
}

/**
 * Build the spawn command for a given tool + directory, accounting for
 * config overrides and custom tool definitions.
 */
export function getSpawnCommand(
  toolId: string,
  directory: string,
  config: WeaveToolsConfig | undefined
): {
  command: string;
  args: string[];
  options: { detached: boolean; stdio: "ignore"; cwd?: string; shell?: boolean; windowsHide?: boolean };
} | null {
  // Defense-in-depth: resolve() normalizes the path (removes ../ segments etc.)
  // and breaks CodeQL's taint chain. The caller (open-directory route) already
  // validates via validateDirectory(), but we sanitize here at point-of-use too.
  const safeDirectory = resolve(directory);

  const platform = process.platform as PlatformId;
  const override = config?.overrides?.[toolId];
  const customDef = config?.custom?.[toolId];

  // Try builtin first
  const builtin = getToolById(toolId);

  if (builtin) {
    const platformCmd = builtin.platforms[platform];
    if (!platformCmd) return null;

    const command = override?.command ?? platformCmd.command;
    const args = override?.args
      ? expandArgsTemplate(override.args, safeDirectory)
      : platformCmd.args(safeDirectory);

    const options: {
      detached: boolean;
      stdio: "ignore";
      cwd?: string;
      shell?: boolean;
      windowsHide?: boolean;
    } = {
      detached: true,
      stdio: "ignore",
    };

    if (platformCmd.options?.shell) options.shell = true;
    if (platformCmd.options?.windowsHide) options.windowsHide = true;
    if (platformCmd.options?.cwd === "directory") options.cwd = safeDirectory;

    return { command, args, options };
  }

  // Try custom tool from config
  if (customDef) {
    if (customDef.platforms && !customDef.platforms.includes(platform)) return null;

    const command = override?.command ?? customDef.command;
    const argsTemplate = override?.args ?? customDef.args ?? "${dir}";
    const args = expandArgsTemplate(argsTemplate, safeDirectory);

    return {
      command,
      args,
      options: {
        detached: true,
        stdio: "ignore",
        shell: platform === "win32" ? true : undefined,
        windowsHide: platform === "win32" ? true : undefined,
      },
    };
  }

  return null;
}

File: src/WeaveFleet.Application/Services/ToolRegistry.cs
using System.Diagnostics;
using System.Runtime.InteropServices;

namespace WeaveFleet.Application.Services;

public sealed record ToolDefinition(
    string Id,
    string Label,
    string IconName,
    string Category, // "editor" | "terminal" | "explorer"
    IReadOnlyDictionary<string, PlatformCommand> Platforms,
    bool AlwaysAvailable = false,
    IReadOnlyDictionary<string, string[]>? DetectBinaries = null);

public sealed record PlatformCommand(
    string Command,
    Func<string, string[]> Args,
    bool Shell = false,
    bool CwdIsDirectory = false);

public sealed record ResolvedTool(string Id, string Label, string IconName, string Category);

public static class ToolRegistry
{
    private static readonly PlatformID CurrentPlatform = Environment.OSVersion.Platform;

    public static string PlatformKey =>
        RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "win32"
        : RuntimeInformation.IsOSPlatform(OSPlatform.OSX) ? "darwin"
        : "linux";

    public static readonly IReadOnlyList<ToolDefinition> BuiltinTools = BuildTools();

    private static readonly Dictionary<string, ToolDefinition> ToolIndex =
        BuiltinTools.ToDictionary(t => t.Id);

    public static ToolDefinition? GetById(string id) =>
        ToolIndex.GetValueOrDefault(id);

    private static List<ToolDefinition> BuildTools()
    {
        return
        [
            // ── Editors ──
            MakeEditor("vscode", "VS Code", "code-2",
                win32: ("code", dir => [dir], shell: true),
                darwin: ("open", dir => ["-a", "Visual Studio Code", dir]),
                linux: ("code", dir => [dir]),
                detectWin: ["code"], detectLinux: ["code"]),

            MakeEditor("vscode-insiders", "VS Code Insiders", "code-2",
                win32: ("code-insiders", dir => [dir], shell: true),
                darwin: ("open", dir => ["-a", "Visual Studio Code - Insiders", dir]),
                linux: ("code-insiders", dir => [dir]),
                detectWin: ["code-insiders"], detectLinux: ["code-insiders"]),

            MakeEditor("cursor", "Cursor", "mouse-pointer-2",
                win32: ("cursor", dir => [dir], shell: true),
                darwin: ("open", dir => ["-a", "Cursor", dir]),
                linux: ("cursor", dir => [dir]),
                detectWin: ["cursor"], detectLinux: ["cursor"]),

            MakeEditor("windsurf", "Windsurf", "wind",
                win32: ("windsurf", dir => [dir], shell: true),
                darwin: ("open", dir => ["-a", "Windsurf", dir]),
                linux: ("windsurf", dir => [dir]),
                detectWin: ["windsurf"], detectLinux: ["windsurf"]),

            MakeEditor("zed", "Zed", "pen-tool",
                darwin: ("open", dir => ["-a", "Zed", dir]),
                linux: ("zed", dir => [dir]),
                detectLinux: ["zed"]),

            MakeEditor("sublime", "Sublime Text", "braces",
                win32: ("subl", dir => [dir], shell: true),
                darwin: ("open", dir => ["-a", "Sublime Text", dir]),
                linux: ("subl", dir => [dir]),
                detectWin: ["subl"], detectLinux: ["subl"]),

            MakeEditor("intellij", "IntelliJ IDEA", "app-window",
                win32: ("idea", dir => [dir], shell: true),
                darwin: ("open", dir => ["-a", "IntelliJ IDEA", dir]),
                linux: ("idea", dir => [dir]),
                detectWin: ["idea", "idea64"], detectLinux: ["idea"]),

            MakeEditor("webstorm", "WebStorm", "app-window",
                win32: ("webstorm", dir => [dir], shell: true),
                darwin: ("open", dir => ["-a", "WebStorm", dir]),
                linux: ("webstorm", dir => [dir]),
                detectWin: ["webstorm", "webstorm64"], detectLinux: ["webstorm"]),

            MakeEditor("rider", "Rider", "app-window",
                win32: ("rider", dir => [dir], shell: true),
                darwin: ("open", dir => ["-a", "Rider", dir]),
                linux: ("rider", dir => [dir]),
                detectWin: ["rider", "rider64"], detectLinux: ["rider"]),

            MakeEditor("goland", "GoLand", "app-window",
                win32: ("goland", dir => [dir], shell: true),
                darwin: ("open", dir => ["-a", "GoLand", dir]),
                linux: ("goland", dir => [dir]),
                detectWin: ["goland", "goland64"], detectLinux: ["goland"]),

            MakeEditor("pycharm", "PyCharm", "app-window",
                win32: ("pycharm", dir => [dir], shell: true),
                darwin: ("open", dir => ["-a", "PyCharm", dir]),
                linux: ("pycharm", dir => [dir]),
                detectWin: ["pycharm", "pycharm64"], detectLinux: ["pycharm"]),

            MakeEditor("rustrover", "RustRover", "app-window",
                win32: ("rustrover", dir => [dir], shell: true),
                darwin: ("open", dir => ["-a", "RustRover", dir]),
                linux: ("rustrover", dir => [dir]),
                detectWin: ["rustrover", "rustrover64"], detectLinux: ["rustrover"]),

            MakeEditor("fleet-jb", "Fleet (JetBrains)", "app-window",
                win32: ("fleet", dir => [dir], shell: true),
                darwin: ("open", dir => ["-a", "Fleet", dir]),
                linux: ("fleet", dir => [dir]),
                detectWin: ["fleet"], detectLinux: ["fleet"]),

            MakeEditor("visual-studio", "Visual Studio", "app-window",
                win32: ("devenv", path => [path], shell: true),
                detectWin: ["devenv"]),

            MakeEditor("xcode", "Xcode", "app-window",
                darwin: ("open", path => ["-a", "Xcode", path]),
                detectDarwin: ["xcodebuild"]),

            MakeEditor("clion", "CLion", "app-window",
                win32: ("clion", dir => [dir], shell: true),
                darwin: ("open", dir => ["-a", "CLion", dir]),
                linux: ("clion", dir => [dir]),
                detectWin: ["clion", "clion64"], detectLinux: ["clion"]),

            MakeEditor("android-studio", "Android Studio", "app-window",
                win32: ("studio", dir => [dir], shell: true),
                darwin: ("open", dir => ["-a", "Android Studio", dir]),
                linux: ("studio", dir => [dir]),
                detectWin: ["studio"], detectLinux: ["studio"]),

            // ── Terminals ──
            new("terminal", "System Terminal", "terminal", "terminal",
                new Dictionary<string, PlatformCommand>
                {
                    ["win32"] = new("cmd", _ => ["/c", "start", "cmd", "/K"], Shell: true, CwdIsDirectory: true),
                    ["darwin"] = new("open", _ => ["-a", "Terminal", "."], CwdIsDirectory: true),
                    ["linux"] = new("x-terminal-emulator", _ => [], CwdIsDirectory: true),
                },
                AlwaysAvailable: true),

            new("wt", "Windows Terminal", "square-terminal", "terminal",
                new Dictionary<string, PlatformCommand>
                {
                    ["win32"] = new("wt", dir => ["-d", dir], Shell: true),
                },
                DetectBinaries: new Dictionary<string, string[]> { ["win32"] = ["wt"] }),

            new("iterm2", "iTerm2", "square-terminal", "terminal",
                new Dictionary<string, PlatformCommand>
                {
                    ["darwin"] = new("open", dir => ["-a", "iTerm", dir]),
                }),

            new("alacritty", "Alacritty", "square-terminal", "terminal",
                new Dictionary<string, PlatformCommand>
                {
                    ["win32"] = new("alacritty", dir => ["--working-directory", dir], Shell: true),
                    ["darwin"] = new("alacritty", dir => ["--working-directory", dir]),
                    ["linux"] = new("alacritty", dir => ["--working-directory", dir]),
                },
                DetectBinaries: new Dictionary<string, string[]>
                {
                    ["win32"] = ["alacritty"], ["darwin"] = ["alacritty"], ["linux"] = ["alacritty"],
                }),

            // ── File Explorers ──
            new("explorer", "File Explorer", "folder-open", "explorer",
                new Dictionary<string, PlatformCommand>
                {
                    ["win32"] = new("explorer", dir => [dir]),
                    ["darwin"] = new("open", dir => [dir]),
                    ["linux"] = new("xdg-open", dir => [dir]),
                },
                AlwaysAvailable: true),
        ];
    }

    private static ToolDefinition MakeEditor(string id, string label, string iconName,
        (string cmd, Func<string, string[]> args, bool shell)? win32 = null,
        (string cmd, Func<string, string[]> args)? darwin = null,
        (string cmd, Func<string, string[]> args)? linux = null,
        string[]? detectWin = null, string[]? detectDarwin = null, string[]? detectLinux = null)
    {
        var platforms = new Dictionary<string, PlatformCommand>();
        if (win32 is var (wCmd, wArgs, wShell))
            platforms["win32"] = new PlatformCommand(wCmd, wArgs, Shell: wShell);
        if (darwin is var (dCmd, dArgs))
            platforms["darwin"] = new PlatformCommand(dCmd, dArgs);
        if (linux is var (lCmd, lArgs))
            platforms["linux"] = new PlatformCommand(lCmd, lArgs);

        Dictionary<string, string[]>? detect = null;
        if (detectWin != null || detectDarwin != null || detectLinux != null)
        {
            detect = new Dictionary<string, string[]>();
            if (detectWin != null) detect["win32"] = detectWin;
            if (detectDarwin != null) detect["darwin"] = detectDarwin;
            if (detectLinux != null) detect["linux"] = detectLinux;
        }

        return new ToolDefinition(id, label, iconName, "editor", platforms, DetectBinaries: detect);
    }

    /// <summary>
    /// Build spawn ProcessStartInfo for a given tool + directory.
    /// </summary>
    public static ProcessStartInfo? GetSpawnInfo(string toolId, string directory)
    {
        var tool = GetById(toolId);
        if (tool is null) return null;

        var platform = PlatformKey;
        if (!tool.Platforms.TryGetValue(platform, out var cmd)) return null;

        var safeDir = Path.GetFullPath(directory);
        var args = cmd.Args(safeDir);

        var psi = new ProcessStartInfo(cmd.Command)
        {
            UseShellExecute = false,
            CreateNoWindow = true,
        };

        foreach (var arg in args)
            psi.ArgumentList.Add(arg);

        if (cmd.Shell && RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
        {
            // On Windows, wrap in cmd /c for shell-based commands
            psi = new ProcessStartInfo("cmd")
            {
                UseShellExecute = false,
                CreateNoWindow = true,
            };
            psi.ArgumentList.Add("/c");
            psi.ArgumentList.Add(cmd.Command);
            foreach (var arg in args)
                psi.ArgumentList.Add(arg);
        }

        if (cmd.CwdIsDirectory)
            psi.WorkingDirectory = safeDir;

        return psi;
    }
}
